### PR TITLE
Form: Add style variations

### LINKED ISSUE #28166: JP Form Block - Adding styles variations to form blocks (-> label placements)
In the third iteration, of improving the [styling of input fields](pcsBup-Gl-p2), we will offer some capabilities for the label placement and behavior by adding form block “styles variations”.

Image

This means we will give users more opinionated options for their input fields, such as an animated label after selecting the input field or placing the label below the input field.

Image

### PR DESCRIPTION
**Pull request by jcheringer**
Fixes #28166

## Proposed changes:
* Includes 2 new Form styles: `Outlined` and `Animated`

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pcsBup-Gl-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:

* Create a post and add a Form block
* Change the form style in the Sidebar to the new styles
* Add some fields and play with the form to ensure everything works as expected

### CODE REVIEW

**Comment by digitalwaveride** [comment]
So much fun, playing with this early draft version! Thanks Jõao!

A few first comments, as currently just the "Outlined" version is testable:

### - Placeholder placement for "Multi-line text input" needs to be "top-left" aligned.

image

### - "Label" & "Required field" are not yet editable

image

### - "Required field" font size is not getting smaller when it animates into the outline (the label does)

image


### - There seems to be some "layer" over the label? Some glitch?

image


### - Placeholder styling: Differences in styling between editor vs published view
Font-size of the placeholder in the editor feels especially off, as the text in the input should be bigger than the label in the outline above.

image

### Padding & Label
Everything needs a bit more "air"... currently there seems to be just a padding of 8px... 

image

... let's increase to all sides to about 16px, and let's change the label font-weight to 300/light, so it looks more like this:

image

### Block select outline in editor

image

Is there anything we can do to avoid this kind of "overlap"?
Similar to when the form wrapper block is selected, and there is some padding to the fields?

image


Really looking forward to where this will go 🚀 ... thanks João!!!

**Comment by jcheringer** [comment]
Thanks for giving it a spin @digitalwaveride!
I didn't face some of the issues you mentioned on my dev environment using Chrome. Maybe some Firefox issues?
I'll investigate them on a JN site.

Regarding the padding, we need to be careful. We can't define an arbitrary number.
We are using the padding defined by the Theme for the input fields.
For the `Twenty Twenty-tree`, I can see the Theme doesn't define paddings for the input fields and uses the styles defined by the Grunion Forms.
In this case, we can touch the Grunion styles to give the fields more room.
Those styles must have a very low specificity so the Themes can override them if necessary.
But, if you look at the Blockbase Theme, you'll see the fields already have a decent spacing:

image

**Comment by digitalwaveride** [comment]
It's getting more and more fun!!! 😄  Thank João! 

## Animated Style

This style works overall already very well! 

#### Multiline Text-Label

In the editor, the padding/line-height needs some finetuning, the first line should look the same as the single line input text-field:

image

#### Dropdown Block 

Styling is still off for the dropdown section... 

image

Also broken in the published view:

image


#### General Label Paddings
We need to increase these. I know, we don't want to overwrite the theme, but we need to do it in a way, where we have a "minimum space" there, independently from the "theme style". 
What I mean is, similar to how we take over the label font-size in the selected state, we need to make sure there is a minimum padding below the label. Label and placeholder text can't be as close as in this theme example. Same for the mc- and single-choice block. Let's have a chat about that.

image



#### Published View

When loading the preview, I see the labels at first without padding, then they jump to the expected place. Happens for me on every page refresh.

https://user-images.githubusercontent.com/5120357/212049274-0a9ead4f-a7bd-4c94-bd79-c320de0262d9.mov

---

## The Below Style

Let's put it that way... there is still a ton of things to figure out ;)
Looking at the finetuning work that is ahead of us with the other two styles, I would suggest putting this style on hold for now, and focus first on the other styles, and then circle back.

--- 

## Outline Style

Played a bit with that style, the published view feels already much better, nice!

In the editor I think there are still a bunch of things open from the previous comment, for giving it later another spin.

Nice work João!

**Comment by jcheringer** [comment]
Hey @digitalwaveride!
I think the PR is ready for another spin. 🙂

**Comment by digitalwaveride** [comment]
Thanks João 🙏 ! Tons of stuff happening here! 🚀 🚀 🚀 

Animated style:
- [x]  **Placeholder placement for "Multi-line text input" needs to be "top-left" aligned.**
- [x]  **"Label" & "Required field" are not yet editable**
- [x]  **"Required field" font size is not getting smaller when it animates into the outline (the label does)**
- [ ]   **Placeholder styling: Differences in styling between editor vs published view**
- [x]  **Dropdown Block - styling the dropdown**

Outlined style:
- [x]  **There seems to be some "layer" over the label? Some glitch?**
- [ ]   **General Padding & Label - Everything needs a bit more "air"**
- [x]  **Block select outline” in editor**
- [ ]   **Glitch when loading published-view → till all styles are loaded**

Regarding the three things that are still open

### Differences in styling between editor vs published view
- Single & multiple choice block, margin between label and answer options are not matching between editor and published view.

image

- Label Styling in Outline Style:

image


- Border styling is also not matching (that said Twenty Twenty Three is a bit "special" in that styling.

- Multiple text-line height is not the same.
image


### Padding / Spacing for Label

Spacing looks good on the outline style, but it's still too dense for the animated style:

Outline:
image

Animated:
image


### Glitch when loading published-view → till all styles are loaded

This still shows, as discussed let's fix this as last item, by making sure all styles are loaded before we show the full form.


João, we are close!!! 😁

**Comment by jcheringer** [comment]
@digitalwaveride, thanks for testing it again.

> Padding / Spacing for Label
Spacing looks good on the outline style, but it's still too dense for the animated style:

What's your suggestion for this one?

> Glitch when loading published-view → till all styles are loaded
This still shows, as discussed let's fix this as last item, by making sure all styles are loaded before we show the full form.

From our discussion, I thought we'd handle this in another PR.
That's not related to these changes and, as you mentioned, there's tons of stuff already happening in this PR.

**Comment by CGastrell** [comment]
I don't have a saying on the styles, I do agree on the "air" thing, but we need to get specific on this and balance how different themes are affected. If we're matching editor/visitor stylings we can address a "better spacing" strategy on a follow up PR.

Same goes for the loading glitch (testing locally is barely noticeable as the styles load quite fast), I think this PR changes a lot of files already and will be troublesome to conciliate with #27837 (whichever lands last). I'd focus on that and then follow up with more changes/fixes

**Review by CGastrell — DISMISSED** [review]
Other than the comment for the follow up, I think this is good to go

**Comment by digitalwaveride** [comment]
>  Padding / Spacing for Label
>     Spacing looks good on the outline style, but it's still too dense for the animated style:
>> What's your suggestion for this one?

I would suggest to aim for the same paddings as for the outline style ... roughly about +8px.

> Glitch when loading published-view → till all styles are loaded
>> From our discussion, I thought we'd handle this in another PR.
👍

**Comment by jcheringer** [comment]
@digitalwaveride I believe I handled the remaining issues.
Whenever you have the time, please, give it another spin.
About adding more space to the Animated style, I've increased the fields padding-top a bit. 
Let me know what you think.
Thanks!

**Comment by digitalwaveride** [comment]
Nearly perfect João! Three little things I noticed:

### Previews
The "previews" in the inspector sidebar when hovering the styling options are not showing yet:

image



### Multi-line Text 

In the published few, for some reason the animation goes not "back", when the field is empty and the field is not selected. See video:


https://user-images.githubusercontent.com/5120357/212960667-4fb327c4-8860-4f8d-a0a5-2da42c51c617.mov



### Dropdown Block glitch is back
Not sure why, but I see the glitch again that you already fixed:

image

**Comment by jcheringer** [comment]
>Previews
The "previews" in the inspector sidebar when hovering the styling options are not showing yet:

I don't know why the previews are not working, and, to be honest, I don't think it's worth spending energy on this issue, at least not on this PR.
The user will see a "live preview" in their form once they change the style.
Also, if you look at the Button block, inside a form, the previews are not working too. 🤷 

> Multi-line Text
In the published few, for some reason the animation goes not "back", when the field is empty and the field is not selected. See video:

Should be fixed now.

> Dropdown Block glitch is back
Not sure why, but I see the glitch again that you already fixed:

I never fixed this issue because it's not happening to me.
Tested on Chrome and Firefox with the Twenty Twenty-three theme.

**Review by digitalwaveride — DISMISSED** [review]
> I don't think it's worth spending energy on this issue, at least not on this PR.

It's a standard for that component, we need to fix it, but fine not to do it in this PR.

And as weird as it is, this time previews show for me, unfortunately though not something useful for the user... it shows the placeholder in the preview, and not any styled input field. 

image

Created an issue draft on our board.

> Multi-line text
✔️ is fixed

> Dropdown Field glitch
Weird, can't reproduce in this version anymore 🤷 


João, let's get this out there! 🚀

**Comment by sdixon194** [comment]
Tested on both Jetpack and Wpcom and a FSE theme - it does seem to work well, however in some instances it seemed to not work consistently. The best I could reproduce was by clicking on a field within the form block -> Clicking on the parent block -> Trying to change the style -> It would return me to the inner block but not change the style. I managed to get a gif of this, but as you  can see it happens kind of sporadically and I haven't 100% figured out what reproduces it yet:

Screen Capture on 2023-01-18 at 11-30-58

>  The "previews" in the inspector sidebar when hovering the styling options are not showing yet:

I would agree that this would be great to get this in as a quick follow up - I wasn't sure what the difference was supposed to be (especially when I was running into the above issue, I was wasn't sure what I was supposed to be seeing until I clicked on it a few times).

**Comment by jcheringer** [comment]
> Tested on both Jetpack and Wpcom and a FSE theme - it does seem to work well, however in some instances it seemed to not work consistently. The best I could reproduce was by clicking on a field within the form block -> Clicking on the parent block -> Trying to change the style -> It would return me to the inner block but not change the style. I managed to get a gif of this, but as you can see it happens kind of sporadically and I haven't 100% figured out what reproduces it yet:

Well, I don't know what I can say or what we can do about that.
For the styles, I'm just using the `style` property in the block's `index.js` file to provide these style variations.
The editor takes care of adding the necessary CSS class, depending on the style selection.
There wasn't an implementation to control that, so it's likely not an issue caused by this PR.
Any insights would be appreciated, though.

>I would agree that this would be great to get this in as a quick follow up - I wasn't sure what the difference was supposed to be (especially when I was running into the above issue, I was wasn't sure what I was supposed to be seeing until I clicked on it a few times).

Regarding the previews, I figured out that by including an `example` in the block definition, the previews will use it.
However, due to the way we are styling the forms ([context](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/modules/contact-form/js/form-styles.js)) , the previews are not styled correctly, and it doesn't seem a trivial change.

I'm counting on your advice on how to proceed with this PR.

@sdixon194 and @digitalwaveride

**Comment by digitalwaveride** [comment]
> Regarding the previews

As discussed above, let's do this in a follow-up PR.

Regarding the issue
> however in some instances it seemed not to work consistently

I also was able to reproduce this. The issue seems to happen only when the switch happens between certain style combinations, like from "Default" style to "Animated". Interestingly, it seems to work reliably if you switch from "Outline" to some other style.

- Change style to "Default"
- Click into a child block, so it's selected.
- Select the parent block.
- Change style to "Animated"
-> no style change happens, instead label of the child block gets selected.

https://user-images.githubusercontent.com/5120357/213695629-d195866b-bbff-4084-b119-119998ed803b.mov

**Comment by jcheringer** [comment]
I remember seeing this issue happening once or twice during the development, but it's not a consistent issue.
Followed the steps you suggested, and I wasn't able to reproduce it on a JN site.

[Screen Recording on 2023-01-20 at 10_16_53.webm](https://user-images.githubusercontent.com/7811225/213704098-b3e8be1f-8d9d-41f0-b39e-9d587a49701b.webm)


But even being able to reproduce the issue, it's like I said, I don't know what we could do.
My implementation isn't responsible for controlling that style state.

**Comment by jcheringer** [comment]
To add a bit more context to the "Switching styles issue":

After a call with @digitalwaveride, we identified that the issue only happens on Firefox.
I spent some time trying to find a cause without success.
Then, I decided to test it on my WordPress.com site, using the command suggested at the beginning of the PR.
It turns out that, on WP.com, the issue doesn't happen. Even on Firefox.
Could that be a Jurassic Ninja issue?

One thing I noticed, but I don't know if it relates to the issue, is that the Sidebar on my dev environment and on WP.com is slightly different from the Sidebar on a JN site.

The Sidebar has Tabs
image

JN Site:
image

Sorry for my lack of knowledge in that aspect, but could that be a Gutenberg version issue?
Right now, I'm trying to guess what could be causing the issue because I don't have a clue about that.

cc @sdixon194

**Comment by sdixon194** [comment]
> It turns out that, on WP.com, the issue doesn't happen. Even on Firefox.

Hmm, I'm able to reproduce this on my wpcom site as well. As you say, looks like it only happens on Firefox. Chrome seems to work fine. 

I wish I had a direction to investigate, but I'm not all that familiar with the block editor I'm afraid. I just happened to notice it when functional testing. 

I'm not sure it's a blocker for the approval from a Crew perspective as long as there's a team review, but would be good to make sure we track it and maybe loop in a team more experienced in working with blocks in the meanwhile.

**Comment by jcheringer** [comment]
Hello folks from @Automattic/jetpack-infinity 👋 
We are having a hard time handling the issue described here: https://github.com/Automattic/jetpack/pull/28212#issuecomment-1398317957
We're not sure if this is something caused by this PR. Our guess is it's not.
Can we have some advice here, please?
Thanks!

**Comment by atanas-dev** [comment]
@jcheringer the Form block is not maintained by Infinity (we're focused on partnerships) so I think @Automattic/jetpack-crew will have a better idea of what the issue is.

File: projects/plugins/super-cache/changelog/super-cache-missing-action [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
```
Comment (jcheringer) [inline projects/plugins/super-cache/changelog/super-cache-missing-action]: Hey @Automattic/jetpack-crew!
I didn't add those super-cache files.
I think this was a side-effect of the merges.
How should I proceed?

**Review by digitalwaveride — DISMISSED** [review]
Nice João!!!
The glitch while switching between styling variations is fixed for me!
🚀

File: projects/plugins/super-cache/changelog/super-cache-missing-action (hunk 1 above)
Comment (jeherve) [reply projects/plugins/super-cache/changelog/super-cache-missing-action]: I'd recommend removing them from your PR. We don't want to add them if they're not necessary.

Comment (sdixon194) [reply projects/plugins/super-cache/changelog/super-cache-missing-action]: Weird, not sure what happened there. Doing some digging, it looks like Super Cache [included those changelogs](https://github.com/Automattic/jetpack/pull/28520/files) in its release the other day, so I would just delete them.

**Review by sdixon194 — COMMENTED** [review]
Yeah! Looks like the issue is fixed for me, nice job! Once those erroneous Super Cache changelog files are deleted, I'll approve and we can get this merged.

**Review by sdixon194 — DISMISSED** [review]

**Comment by donnchawp** [comment]
I noticed a few problems with the dropdown input when the border-radius on an input field is set, in the Outlined style.

Screenshot 2023-01-30 at 11 58 35

- [ ] The label causes the border to bulge at the top.

Screenshot 2023-01-30 at 12 40 31

- [ ] The background colour of the field  is under the label text.

Screenshot 2023-01-30 at 11 59 25

- [ ] When there are more than 3 options, the right hand side of the enclosing border has a different border-radius to show the vertical scroll bar. That does happen in trunk on the frontend using Twenty Twenty-One, but in the editor you just see the border-radius as normal on each side of the dropdown field.

Screenshot 2023-01-30 at 12 54 44

- [ ] In Twenty Twenty-One/Two/Threee the dropdown input field borders are messed up by the label on the frontend. I didn't test any other themes.

**Comment by donnchawp** [comment]
Again in Outlined style, the label of the input field bounces up and down when adjusting the border width. It happens in any of the input fields that have an adjustable border, but here's a dropdown field.
Would it be possible to lock the location of the label while the border is being adjusted?

https://user-images.githubusercontent.com/5656673/215487508-b25fd766-7550-4037-94ec-0399b1bc6407.mov

**Comment by digitalwaveride** [comment]
There seems to be a new glitch in the outline style with the border in the editor (testing on Firefox): 

image

Not visible in the published view:
image

**Comment by donnchawp** [comment]
Screenshot 2023-01-30 at 13 36 20

- [ ] The Email field warps the border too.

**Comment by digitalwaveride** [comment]
João, thanks so much for all the fixes, and adjustments! 🚀 
Things tested well for me!

Besides the "preview on styles hover" thing, that we wanted to fix in a separate PR, I found one more little thing, hopefully not a big issue:

The Multiple Choice and Single Choice block seem to have a block background assigned.
As we decided that we hide the background color styling, the block should better behave like the checkbox block, with a transparent background. That way the user has a chance to add via a group block etc. their own background if needed.
 
image

**Comment by jcheringer** [comment]
> João, thanks so much for all the fixes, and adjustments! 🚀 Things tested well for me!
> 
> Besides the "preview on styles hover" thing, that we wanted to fix in a separate PR, I found one more little thing, hopefully not a big issue:
> 
> The Multiple Choice and Single Choice block seem to have a block background assigned. As we decided that we hide the background color styling, the block should better behave like the checkbox block, with a transparent background. That way the user has a chance to add via a group block etc. their own background if needed.
> 

I'm not sure if that is the correct way to fix the 'issue'.
If I remove the background from the Multiple and Single Choice fields, that's what you'd see on a theme like Twenty Twenty-one:

image

In the Animated and Outlined style, we added the same background color from the input fields to match the 'boxed' form style:

image

In my opinion, the correct way to fix this is to add the possibility to change the colors and the border settings like we did for the other fields. But that's something for another PR.

**Review by digitalwaveride — COMMENTED** [review]
>  way to fix this is to add the possibility to change the colors and the border settings like we did for the other fields. But that's something for another PR.

Agreed! As discussed in our call today, let's do this in a follow-up PR that follows up shortly. 
Thanks João! 👍

**Review by digitalwaveride — DISMISSED** [review]

**Review by jeherve — COMMENTED** [review]
This tests well for me in Jetpack, and in the post editor on WordPress.com Simple.

In the site editor, however, the blurring effect stays there whenever I add a form to the site editor:

image

This doesn't seem to happen on `trunk`, so this may be caused by this PR. Does this happen to you too?

**Comment by jcheringer** [comment]
> This tests well for me in Jetpack, and in the post editor on WordPress.com Simple.
> 
> In the site editor, however, the blurring effect stays there whenever I add a form to the site editor:
> This doesn't seem to happen on `trunk`, so this may be caused by this PR. Does this happen to you too?

Do you mean the "Editor" under the "Appearance" menu?
image

If so, I tested it on my WP site, and I noticed something weird.
After adding the Form block to the page, the blur effect persisted until I moved my mouse over the form.
But then I cleaned the page up by removing most of the content, and now I'm not able to reproduce it anymore.

[Screen Recording on 2023-02-02 at 16_21_30.webm](https://user-images.githubusercontent.com/7811225/216429358-73f5057b-83fd-4029-b9fb-7f0d41c850ce.webm)

**Comment by jeherve** [comment]
> Do you mean the "Editor" under the "Appearance" menu?

Yep, that's the one. 👍

> I tested it on my WP site, and I noticed something weird.

Yes, that blur is problematic. Here is what I'm seeing on my end on WordPress.com Simple. I do not seem to have that issue without that branch applied:

https://user-images.githubusercontent.com/426388/216590990-c63f32af-5ef0-4d48-8040-07e3030b507a.mov

**Comment by jcheringer** [comment]
Thanks for sharing the video @jeherve!
I couldn't reproduce it yet, but I don't want to hold this PR anymore because of that.
The blur effect was supposed to improve things a bit on the user side, but we can get back to it in a follow-up PR.
For now, I've removed the blur effect from the Editor.

cc @digitalwaveride

**Review by jeherve — APPROVED** [review]

**Pull request merged**